// firewire/index.js
var Firebase = require("firebase");

exports.wire = function (req,res,fw, config) {
	fw.ref = new Firebase(fw.url);

	fw.children=[];
	fw.childrenReturned=[];

	fw.server = {
		f:{},
		map:{}
	};

	var x,y,z,fbUrl, fChild;

	var firebaser = function(fbUrl,fChild){
		
		fw.ref.child(fbUrl).on("value", function(snapshot) {
			fw.server.f[fChild] = snapshot.val();
			fw.server.map[fChild] = fbUrl;
			fw.childrenReturned.push(fChild);
			if(fw.children.length === config.length) {
				if(fw.children.length === fw.childrenReturned.length){
					console.log("firewire: all data resolved | loading", req.params.page);
					res.render(req.params.page, {
						title : "Welcome to firewire",
						root : req.protocol + '://' + req.get('host'),
						firewire : JSON.stringify(fw.server),
						fw : fw.server.f,
						pageLoad : req.params.page
					});
				}
			}
		});
	};

	var cycleThroughData = function(){
		//Cycle through config array
		for (x in config){
			//For each obj in array get the firebase url and the child of f to pass the data to...
			for (y in config[x]){
				fbUrl = y;
				fChild = config[x][y];

				for (z in req.params){
					fbUrl = fbUrl.replace(":" + z , req.params[z]);
				}

				fw.children.push(fChild);
				firebaser(fbUrl,fChild,req);
			}
		}
	};

	if (config.length === 0){
		res.render(req.params.page, {
			title : "Welcome to firewire",
			fw : {}
		});
	}
	else {
		cycleThroughData();
	}
	
};
